### ISSUE TITLE
Transposition operators

### ISSUE THREAD

**Issue by daveyarwood** [alda-lang/alda#129]
Per @MillerMark in #7:

(EDIT: see comments below for further development of syntax ideas)

> Add two more built-in operators for pitch shifting the entire group. The multiplication operator and the pitch-shift operators would behave consistently with the individual note pitch shift operators (examples below).
> 
> ```
> # pitch shift up operator (shifts up semi-tones)
> [c d e f g]+
> [c d e f g]+3
> melody+5
> upDownInF = [upDownInC+5]
> upDownInG = [upDownInF++]
> 
> # these two are equivalent:
> [c d e f g]++++
> [c d e f g]+4
> 
> # pitch shift down operator (shifts down semi-tones)
> melody-5
> 
> # these two are equivalent:
> [ c d e f g ]--
> [ c d e f g ]-2
> 
> # operators can be combined (order doesn't matter - same results).
> [ c d e f g ]+2*2
> melody*2-2
> 
> # these two are equivalent:
> melody*3+4
> melody***++++
> ```

and:

> Pitch modifiers ("+" and "-") optionally follow a group or variable and adjust the pitch of all the notes in that variable or note group. Pitch-shift modifiers can be followed by a number (e.g., "-7"), or the "+" and "-" operators can be repeated in any combination for as many semitones you need to adjust (just like Alda's syntax for flat/sharp semitone adjustment).
> 
> Examples:
> 
> ```
> crun = o3 c4 e g a a+ a g e
> frun = crun+5
> grun = frun++
> gChord = o2 g4/b/>d r64 <g1~1~1/b/>d
> fChord = gChord--
> cChord = fChord-5
> ```

Brilliant idea, IMO. Probably a common enough use case to merit inclusion in the language itself (vs. available as a plugin), though that's open for discussion. Thoughts?

**Comment by daveyarwood** [comment]
Should allow transposition to be set as an attribute as well, per @elyisgreat's suggestion.

**Comment by elyisgreat** [comment]
Any progress? By the way, I'd imagine that both forms of transposition would support decimals for use with microtones, although I think Hz frequencies should have their own notation.

**Comment by daveyarwood** [comment]
I haven't had time to work on this feature -- my current focus is preparing Alda for a [1.0.0 release](https://github.com/alda-lang/alda/milestones/1.0.0). I'm definitely on board with this idea though. A PR would be great, if anyone would like to take a stab at it.

Being able to transpose by a frequency in Hz would be a nice feature as well. I'm not sure yet what that syntax would look like. Maybe it would look like `[c d e f g]+1000hz`, or something like that.

**Comment by elyisgreat** [comment]
I disagree with the transposition by Hz, as transposition is a fundamentally multiplicative, not additive operation. Adding 1000Hz to 4000Hz creates a different interval than adding 1000Hz to 2000Hz. It would be nice to multiply the frequencies direct, however since transposing by n semitones is multiplying the frequency by (2^(1/12))^n, you can multiply a frequency by x by transposing by log(x)/log(2^(1/12)) semitones.

**Comment by daveyarwood** [comment]
Oh yeah, you're right. Simply adding 1000 Hz to a section of notes would effectively make them all sound out of tune with each other :stuck_out_tongue: 

Perhaps it would be more useful to allow transposing by fractions of a semitone?  Like `[c d e f]+0.5` would result in tuning those notes 1/4 step sharp.

**Comment by elyisgreat** [comment]
Yeah. I kinda assumed fractional transposition would be supported.

**Comment by daveyarwood** [comment]
I'm on board with that. We should start with semitone transposition as a first step, and we can then add fractional transposition later -- that part will depend on #87.

**Comment by elyisgreat** [comment]
Oh. Also, since `+n` syntax might create conflicts, like `d8 f+4`, where it could either be an eighth-note A or a quarter-note F\#, why not have the syntax `^n`?

**Comment by daveyarwood** [comment]
Hmm, good point. I like the "+" syntax, though, and I think transposing a
single note isn't very useful. So, we could just restrict this to event
sequences.

**Comment by elyisgreat** [comment]
I think even with restricting the `+` syntax to event sequences it's still confusing. Why not allow `+` and `-` to increment/decrement event sequences by one semitone and `^n` to increment notes and event sequences by n semitones?

**Comment by daveyarwood** [comment]
Having a special operator (^) for transposition by semitones could be
valuable, especially if we supported fractions of semitones and negative
numbers. Then you could tune individual notes like:

c4^-0.5  (C quarter note, detuned half a semitone)
d+^4 (D sharp, plus 4 semitones, so it sounds like G)

**Comment by elyisgreat** [comment]
Agreed. Then we could have the two modes for transposition for both notes and event sequences, as well as an attribute.

**Comment by daveyarwood** [comment]
Yep, exactly. I like it!

**Comment by elyisgreat** [comment]
Detailed from #236 

Currently, the `+` and `-` operators only apply to the event that they're connected to. However, according to the plans we currently have, transposition can act as both an operator or an attribute. But unlike operators, attributes apply to the entire instrument at the time they are called until another attribute call is made. If transposition is to be both an operator and an attribute, this could result in weird behaviour where the operators act like attributes. For example, take this line: `(transpose 3) g g^2 g`. If the operators are to behave like attributes, this would result in the weird quirk where the third g would sound like c, as opposed to b-flat, as if the ^2 hadn't been applied at all. And what if the code were `g+ g`? This would result in the equivalent of 2 g-sharps as opposed to a g-sharp followed by a g.

**Comment by daveyarwood** [comment]
What we could do is make the transposition _operator_ only act on the current note. In that sense it would be kind of like the flat/sharp operators, except that you can specify how many half-steps you want to move up/down, without affecting subsequent notes. To make it "sticky," the `transpose` _attribute_ could be used, and all subsequent notes would be affected.

**Comment by elyisgreat** [comment]
Yeah. It's slightly confusing, but then again, "sharp/flat" was never an attribute, so one could argue that this the operator an extension of "sharp/flat" while *transpose* is an entirely new attribute. 

Another thing: I noticed in the parsed clojure code that the sharps were marked by `:sharp`. An improvement would be to store a "transposition number" instead of the `:sharp` markers. And as for operating on event sequences, I would imagine that it would simply add the transposition number on the event sequence to the number of every note inside. 

This will also increase the character limit until #235 is fixed. 

And just in case your worried about the math of all these new attributes, I've worked it out. The frequency of any given note would be 

(tuningConstant)\*2^(transpose/12)\*2^(octave-4)\*2^(noteValue)\*2^(transpositionNumber/12)

(noteValue is a constant depending on whether the note is a c, d, e, f, g, a, or b, in which case it would be -9/12,-7/12,-5/12,-4/12,-2/12,0/12, and 2/12, respectively.)
